### PR TITLE
feat(wallet): add ability to send a token

### LINKED ISSUE #18005: Integrate endpoints to send transaction
# Problem

Currently we have UI in place with mocked data for sending transactions, but transactions are not executed. We need to integrate with corresponding status-go endpoints for sending transactions

### PR DESCRIPTION
**Pull request by J-Son89**
fixes: https://github.com/status-im/status-mobile/issues/18005

This pr adds the integration with Status Go to send a transaction.

Testing Notes:
This pr is adding the happy path for sending a token. 
To send a token this is the flow. 
- On test net or prod
- go to wallet
- go to an account, hit "send" button.
- enter an address using the input box or the scanner (don't select your own accounts via the my-accounts tab)
- choose an amount in crypto less than the amount in your account to allow for gas etc.
- continue to confirmation page
- authorize and sign the transaction
- should then navigate to transaction progress page

This is only for the mainnet transactions (including testnet), support for other Optimism & Arbitrum will be added in addtional pr's.

* Transaction Confirmation page is missing a few UI details, such as gas, time of transaction etc. I will create a follow up issue to implement these completely. 

* Transaction progress page is only initial state of it, there is no expected functionality there as this will be handled in a separate issue. For now the button on that page will navigate to the account-page.
This navigation should also end up with the Transactions tab active on the send page, this is beyond the scope of this pr. The main focus is to be able to send any token 

The slide button on the transaction confirmation page is not sitting right. This is a common component and I would prefer to handle this issue as I don't want to block this pr because of that. The send flows are high priority for 1.27 and this issue can be handled in isolation 👍 

Design Review:
Quo Component - Tags -> Summary Tag was also revisited as it was missng some designs. It would be good to check this component is looking alright here 👍 

https://github.com/status-im/status-mobile/assets/22799766/1248954a-4be4-4e5d-8c3d-fa1b0b5775e6

### CODE REVIEW

**Comment by J-Son89** [comment]
I can't figure out what it is, but the floating button page is not sitting right for slide button.



If I hot reload it sits perfectly though 😅 



If anyone has any ideas what the fix is would be appreciated

File: src/quo/components/wallet/summary_info/view.cljs
Comment (J-Son89) [inline src/quo/components/wallet/summary_info/view.cljs:42]: the designs only show these if they have a value. We could also look to remove network specific details out of this component and make it more generic but it's for another pr as this work is already growing too large.

File: src/quo/components/wallet/summary_info/view.cljs
Comment (J-Son89) [inline src/quo/components/wallet/summary_info/view.cljs]: some slight differences for designs then what was initially implemented here. I think it went unnoticed and will ask for a design review for this component.

File: src/status_im/common/standard_authentication/standard_auth/slide_button/view.cljs
Comment (J-Son89) [inline src/status_im/common/standard_authentication/standard_auth/slide_button/view.cljs]: will adjust formatting 🤒

File: src/quo/components/wallet/summary_info/view.cljs
Comment (J-Son89) [inline src/quo/components/wallet/summary_info/view.cljs]: will adjust formatting 🤒

File: src/utils/money.cljs
Comment (J-Son89) [inline src/utils/money.cljs:250]: not sure about this, but the money utils are quite generic and `amount-in-hex` seems more specific to the wallet. So I moved it to utils there, but happy to leave it as is if preferred 👍

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs:131]: max fees to be calculated properly in a follow up 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (OmarBasem) [inline src/status_im/contexts/wallet/send/events.cljs]: What is that hardcoded "0x" ?

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (OmarBasem) [inline src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs]: Can the keys in this data be converted to kebab-case `(:EstimatedTime route)`

File: src/status_im/contexts/wallet/send/events.cljs
Comment (OmarBasem) [inline src/status_im/contexts/wallet/send/events.cljs:139]: Many keys here are not kebab-case

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs]: yeah you're right, will adjust these!

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs]: I will find the explanation and send it on. For now this is okay to leave as is though.

File: src/quo/components/wallet/summary_info/view.cljs
Comment (ulisesmac) [inline src/quo/components/wallet/summary_info/view.cljs:39]: Do we have a place to extract these strings ("ETH", "OPT", etc) by network name (e.g. :optimism) ?

File: src/quo/components/wallet/summary_info/view.cljs
Comment (ulisesmac) [inline src/quo/components/wallet/summary_info/view.cljs]: It can be a `case`

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:59]: :+1:

File: src/status_im/contexts/wallet/send/utils.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/utils.cljs:6]: Maybe we should get some tests for this function?

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:106]: :eyes:

File: src/utils/money.cljs
Comment (ulisesmac) [reply src/utils/money.cljs:250]: It has been moved to `status-im.contexts.wallet.send.utils` maybe it should have been moved to `status-im.contexts.wallet.utils` instead?

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs]: This event looks a little large. I think we could split it into more small pieces.

Most of the problems related to why it is verbose are related to CamelCase, so I'd suggest to transform the keys keys while being stored:

```clojure
(rf/reg-event-fx
 :wallet/suggested-routes-success
 (fn [{:keys [db]} [suggested-routes timestamp]]
   (when (= (get-in db [:wallet :ui :send :suggested-routes-call-timestamp]) timestamp)
     {:db (-> db
              (assoc-in [:wallet :ui :send :suggested-routes] suggested-routes)
              ;; Transform the keys in `(first (:Best suggested-routes))` to kebab case
              (assoc-in [:wallet :ui :send :route] (first (:Best suggested-routes)))
              (assoc-in [:wallet :ui :send :loading-suggested-routes?] false))})))
```
That will also solve the problem @OmarBasem said about transforming the estimated time route.

Once that is done, we could refactor this event easier, just as a suggestion:
```clojure
(rf/reg-event-fx
 :wallet/send-transaction
 (fn [{{wallet-db :wallet} :db} [sha3-pwd]]
   (let [{:keys [route token
                 to-address]}      (get-in wallet-db [:ui :send])
         from-address              (:current-viewing-account-address wallet-db)
         token-id                  (:symbol token)
         multi-transaction-command {:fromAddress from-address
                                    :toAddress   to-address
                                    :fromAsset   token-id
                                    :toAsset     token-id
                                    :fromAmount  (:amount-out route)
                                    :type        0}
         transaction-bridge        (get-transaction-bridge {:from-address from-address
                                                            :to-address   to-address
                                                            :route        route})
         request-params            [multi-transaction-command transaction-bridge sha3-pwd]]
     {:json-rpc/call [,,,]})))
```

That is getting the transaction bridge map in a separate function. I'm not sure rn, but I think getting a transaction bridge will be a common operation, so maybe that function can be reused or extended in the future. ATM, it just helps to readability, that function is:
```clojure
(defn- get-transaction-bridge [{:keys [from-address to-address route]}]
  (let [{:keys [from bridge-name amount-out gas-amount gas-fees]}           route
        {:keys [gas-price max-fee-per-gas-medium max-priority-fee-per-gas]} gas-fees]
    [{:BridgeName bridge-name
      :ChainID    (:chain-id from)
      :TransferTx {:From                 from-address
                   :To                   to-address
                   :Gas                  (money/to-hex gas-amount)
                   :GasPrice             (money/to-hex (money/->wei :gwei gas-price))
                   :Value                amount-out
                   :Nonce                nil
                   :MaxFeePerGas         (money/to-hex (money/->wei :gwei max-fee-per-gas-medium))
                   :MaxPriorityFeePerGas (money/to-hex (money/->wei :gwei max-priority-fee-per-gas))
                   :Input                ""
                   :Data                 "0x"}}]))
```

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:105]: Why do we only get the first item? what is contained in the rest of them?

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs]: Are we setting the remaining networks as zero because we are not supporting them rn?
If so, maybe we can remove them or add a comment specifying this is temporal.

**Review by ulisesmac — COMMENTED** [review]
Code looks good @J-Son89 I just left some suggestions

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:109]: Just a question here,

Are transactions associated to an account?
If so, please consider adding these transactions inside the respective account

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs]: Data is a secondary feature, it is something which can be useful when making something other than an ETH transaction. It is the encoding of the contract call. We will eventually follow up to include this feature.
cc @alaibe in case I left out any details

File: src/quo/components/wallet/summary_info/view.cljs
Comment (J-Son89) [reply src/quo/components/wallet/summary_info/view.cljs:39]: I would like to revisit how this component is working - I will create a follow up issue for this. For this pr it goes outside of the scope and send already is a huge feature set. imo the quo components should have no knowledge of networks and should just render the UI.

File: src/status_im/contexts/wallet/send/utils.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/utils.cljs:6]: sure, we can do this. 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:106]: will adjust the naming here 😅

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs]: Yep, I agree with everything here and about the keys neededing to be adjusted. Will do all! Thanks for the feedback @ulisesmac  & @OmarBasem  :)

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:105]: ah this is a temporary solution, will adjust properly when we go to implement the transaction progress page. 
Happy to remove it for now, I think I left it there to also open up discussion about how we store transactions in the db

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs]: I need to remove this and this should be pulled from data 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:109]: let's discuss this today!

File: src/utils/money.cljs
Comment (J-Son89) [reply src/utils/money.cljs:250]: Send is such a large area I would prefer to keep some utils there that are specific to send related parts!

Comment (J-Son89) [reply src/utils/money.cljs:250]: I also prefer to have some naming to utils once they grow too much, not a big fan of lots of random functionality in a utils file, we should also keep an eye on that imo

File: src/quo/components/utilities/token/loader.cljs
Comment (J-Son89) [inline src/quo/components/utilities/token/loader.cljs]: @ulisesmac and I revisted this as we realised it was breaking the token component 👍

File: src/quo/components/utilities/token/view.cljs
Comment (J-Son89) [inline src/quo/components/utilities/token/view.cljs:42]: @ulisesmac & I found a bug where empty tokens didn't sit right, this fix that 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs:14]: will delete empty line 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs]: lowered routes keys 👍

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs]: put it in a small helper function 👍

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs]: done

File: src/quo/components/wallet/summary_info/view.cljs
Comment (J-Son89) [reply src/quo/components/wallet/summary_info/view.cljs]: done!

**Comment by J-Son89** [comment]
Would appreciate another review / if anyone wants to test this too would be great ! :)

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:105]: Discussed with @ulisesmac - will create an issue for this feature and mention this needs to be reworked. 👍

**Review by OmarBasem — APPROVED** [review]
Looks good 👍

File: src/status_im/contexts/wallet/send/utils.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/utils.cljs:6]: forgot to add this, doing now!

Comment (J-Son89) [reply src/status_im/contexts/wallet/send/utils.cljs:6]: Done

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs:143]: Nit: a `prn` statement :)

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (smohamedjavid) [inline src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs:167]: Could we use fragment `[:<>]` here, as we don't have any styles?

**Review by smohamedjavid — APPROVED** [review]
Great work! @J-Son89 🚀

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs:143]: Thanks!

File: src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/transaction_confirmation/view.cljs:167]: Will check 👌

File: src/status_im/contexts/wallet/send/events.cljs
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:140]: Really nice :+1: :100:

**Review by ulisesmac — APPROVED** [review]
Good work! :100:

File: src/status_im/contexts/wallet/send/events.cljs
Comment (briansztamfater) [inline src/status_im/contexts/wallet/send/events.cljs:60]: 😎

**Review by briansztamfater — APPROVED** [review]
🚀

File: translations/en.json
Comment (vkjr) [inline translations/en.json:2449]: A similar label already exists in `en.json`:
`"to-capitalized": "To",`

**Review by vkjr — APPROVED** [review]

File: src/status_im/contexts/wallet/send/events.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/events.cljs:105]: https://github.com/status-im/status-mobile/issues/18307

**Comment by pavloburykh** [comment]
@J-Son89 thank you for the PR. 

Speaking about happy flow - I was able to send some amount of EHT from Account A to Account B. 

Besides there are plenty issues, you have already mentioned some of them. I don't know if it makes sense describe all the issues, as you have pointed out that this flow is still in progress. I will mention some of them below so you can keep an eye on them in the follow up PRs. 

### ISSUE 1 Not able to confirm sending STT (Status Test Token). Confirm button just does not respond.

https://github.com/status-im/status-mobile/assets/97245802/f0c15135-766e-46d5-9588-3a44b5388044

### ISSUE 2 Unable to return to wallet screen  after completing the transaction 

**On IOS** at the end user appears on the empty wallet screen. Need to relogin the app to unblock navigation to other screens. 

**IOS Video**

https://github.com/status-im/status-mobile/assets/97245802/f865f7b8-28eb-468a-83b4-ea75ac302c81

**On Android** not able to close Sending screen

**Android video:**


https://github.com/status-im/status-mobile/assets/97245802/c072be20-b7a2-450d-b068-5e922f835f74


### ISSUE 3 Balance is not updated for token receiver until relogin

### ISSUE 4 Balance is rounded to 2 decimals on “Select asset” screen. 

photo_2024-01-04 18 14 21
photo_2024-01-04 18 14 24


### ISSUE 5  “Doesn't support name” error  when selecting ETH token on “Select asset” screen. 

photo_2024-01-04 18 26 36


Cannot reproduce constantly. I have faced it only couple of times so far. 

[Status-debug-logs (93).zip](https://github.com/status-im/status-mobile/files/13832636/Status-debug-logs.93.zip)

**Comment by pavloburykh** [comment]
> Design Review:
> Quo Component - Tags -> Summary Tag was also revisited as it was missng some designs. It would be good to check this component is looking alright here

@J-Son89 do I understand correctly that this PR needs design review based on this comment?

**Comment by J-Son89** [comment]
@pavloburykh - thanks for all the reviews and mentioning those issues. I will check all these and create the issues as needed!
Yes it would be great if @Francesca-G could take a look at the Summary component - (quo tags/ summary-tag) and do a design review of that. It was missing some variants which I have added here.
However it might be nice to do that and track any issues as a follow up as the component adjustments was beyond the initial intention of this pr. wdyt? 🤔 

and also just to note that particular component is not used any where else in the codebase (except for the preview screen)

**Comment by pavloburykh** [comment]
@J-Son89 thanks! So let's wait for Francesca's review. Meanwhile I have triggered e2e re-run as there were some failures in the previous run (most of them are 100% PR related though). I will provide feedback when the results are ready. Based on the e2e runs queue the results will be ready in couple of hours, so hopefully will be ready to merge the PR tomorrow. 

Sorry once again for the delay with testing Jamie.

**Comment by J-Son89** [comment]
Not at all, thanks for the rigorous testing! Really helps put the feature together 🙏

**Comment by pavloburykh** [comment]
@J-Son89 failed e2e are not PR related. So we are ready to merge after design review.

**Review by Francesca-G — COMMENTED** [review]
> Design Review:
Quo Component - Tags -> Summary Tag was also revisited as it was missng some designs. It would be good to check this component is looking alright here 👍

Here's the design review for the Summary Tag component :)
https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS/Feedback-for-Mobile?type=design&node-id=3415%3A214763&mode=design&t=ANYgMF9xQ4Gp2lOW-1

**Comment by J-Son89** [comment]
Thanks for the review @Francesca-G - I will create a separate issue to make those adjustments. 🙏

**Comment by J-Son89** [comment]
@pavloburykh - I tracked these issues:
Issue 1: https://github.com/status-im/status-mobile/issues/18413

Issue 2 - will be done as part of 
https://github.com/status-im/status-mobile/issues/18307

Issue 3: https://github.com/status-im/status-mobile/issues/18412

Issue 4: https://github.com/status-im/status-mobile/issues/18414

Issue 5-
https://github.com/status-im/status-mobile/issues/18411

@Francesca-G - design review updates tracked here:
https://github.com/status-im/status-mobile/issues/18410

**Comment by mariia-skrypnyk** [comment]
Hi @J-Son89 !

Have some update for **ISSUE 1 Not able to confirm sending STT (Status Test Token). Confirm button just does not respond.** 

Now we have the same situation with ETH at newest nightly build from the 10 of January. So we have no ability to finish a transaction. 

Also wanted to ask if you had a chance to create issues mentioned here?  

> @pavloburykh - thanks for all the reviews and mentioning those issues. I will check all these and create the issues as needed!

**Pull request merged**
